### DISCUSSION TITLE
One-test files

### DISCUSSION THREAD

**Discussion by pwntester** [nvim-neotest/neotest#420] [General]
Hi,

I was considering writing an adapter for CodeQL tests. Theses tests are defined in an non-standard way though, there are multiple directories and each of them contains a `.ql` or `.qlref` file which contains the query to run as a test. So each file is itself a single query. Considering that neotests expects a file to be parsed to extract namespaces and tests from it, would it fit this other way of defining tests? What should `Adapter.discover_positions(path)` return? We cant use treesitter here since `.qlref` files just contain the location of a query to be run as part of a test. Can we build a `Neotest.Tree` with a single node as if there was a single position in each test file? Any adapters working in a similar way that I could use as a reference?

Thanks!

**Comment by rcarriga** [comment]
Hey great to hear you want to write an adapter! :smile: Unless I'm misunderstanding, this should be relatively simple since it's just a single node and neotest doesn't require that you have children in the file. 

Something like this should be sufficient
```lua
local lib = require("neotest.lib")
local Tree = require("neotest.types").Tree

local Adapter = {}

---@param file_path string Absolute file path
---@return neotest.Tree | nil
function Adapter.discover_positions(file_path)
  local sep = lib.files.sep
  local path_elems = vim.split(file_path, sep, { plain = true })
  return Tree.from_list({
    {
      type = "file",
      path = file_path,
      name = path_elems[#path_elems],
    },
  })
end
```
Though I might be missing something as I haven't tested.

Let me know if you have any issues!

**Reply by pwntester** [reply]
Thanks! Will give it a go and hopefull submit a new adapter soon!

**Reply by pwntester** [reply]
@rcarriga the way that codeql tests work they need to get a lock on a shared database so running tests concurrently will make them fail if they belong to the same directory. I can process all the tests to be run and reduce them to a set of commands that can safely be run concurrently. The problem I have is that when marking several tests in the summary panel and running them all, it triggers several calls to build_spec so I dont get an holistic vision of all the tests that are going to be executed and process them into safe commands to be run. Is there any way to hook into the RunSpecs to be executed and modify them before they start spawning processes?
